File: apps/api/src/execution/newsBrief.ts
/**
 * Morning news brief → Telegram. Part of the agent's "check news first" daily
 * routine: before the session, summarize the high/medium-impact economic events
 * in the next 24h and push them to the user's Telegram chat.
 *
 * Best-effort throughout: a missing AI service, empty calendar, or unconfigured
 * Telegram must never throw into the scheduler. The same NewsEvent rows feed the
 * risk engine's ±news-window block, so this brief and the trade gate stay in sync.
 */

import { prisma } from "../lib/prisma";
import { defaultChatId, isConfigured, sendMessage } from "../telegram/telegram";
import { collectMarketContext } from "./dailyBriefing";

const DAY_MS = 24 * 60 * 60 * 1000;
const AI_SERVICE_URL = process.env.AI_SERVICE_URL ?? "http://localhost:8000";

/** Escape the few characters Telegram's HTML parse_mode treats specially. */
function esc(s: string): string {
  return s.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;");
}

interface BriefEvent {
  title: string;
  impact: string;
  currency: string;
  scheduledAt: Date;
  forecast: string | null;
  previous: string | null;
}

/**
 * Ask the AI service for a short consolidated "what to watch" paragraph. Returns
 * null on any failure so the brief still goes out with just the event list.
 */
async function aiSummary(events: BriefEvent[]): Promise<string | null> {
  try {
    const headlines = events.slice(0, 25).map((e) => ({
      title: e.title,
      source: "economic-calendar",
      publishedAt: e.scheduledAt.toISOString(),
      body: `${e.impact}-impact ${e.currency} event. Forecast: ${e.forecast ?? "n/a"}, Previous: ${e.previous ?? "n/a"}.`,
    }));
    const res = await fetch(`${AI_SERVICE_URL}/analyze/news-summary`, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({ headlines }),
    });
    if (!res.ok) return null;
    const json = (await res.json()) as { summary?: string };
    return json.summary?.trim() || null;
  } catch {
    return null;
  }
}

/** Morning desk call from the market-context agent: bias + levels + risks per
 *  traded series. Empty string when the AI service has nothing to say. */
async function deskCallSection(): Promise<string[]> {
  const contexts = await collectMarketContext();
  if (contexts.length === 0) return [];
  const lines = ["", "<b>DESK CALL</b>"];
  for (const m of contexts) {
    const dot = m.bias === "Bullish" ? "🟢" : m.bias === "Bearish" ? "🔴" : "⚪";
    lines.push(`${dot} <b>${esc(m.symbol)} ${esc(m.timeframe)}</b> — ${esc(m.bias)}`);
    lines.push(esc(m.summary.slice(0, 300)));
    if (m.keyLevels.length) lines.push(`Levels: ${esc(m.keyLevels.slice(0, 3).join(" · "))}`);
    if (m.risks.length) lines.push(`Risks: ${esc(m.risks.slice(0, 2).join(" · "))}`);
  }
  return lines;
}

/** Build the HTML message body for the next-24h news brief. */
export async function buildNewsBrief(now: Date = new Date()): Promise<string> {
  const events = await prisma.newsEvent.findMany({
    where: {
      scheduledAt: { gt: now, lt: new Date(now.getTime() + DAY_MS) },
      impact: { in: ["HIGH", "MEDIUM"] },
    },
    orderBy: { scheduledAt: "asc" },
    take: 15,
  });

  const desk = await deskCallSection();

  if (events.length === 0) {
    return [
      "📰 <b>NEWS BRIEF</b> — next 24h (UTC)",
      "",
      "No high- or medium-impact events scheduled. Clear to follow your plan — still read the chart before any entry.",
      ...desk,
    ].join("\n");
  }

  const summary = await aiSummary(events);
  const lines = ["📰 <b>NEWS BRIEF</b> — next 24h (UTC)", ""];
  if (summary) lines.push(esc(summary), "");

  lines.push("<b>SCHEDULED</b>");
  for (const e of events) {
    const time = e.scheduledAt.toISOString().slice(11, 16); // HH:MM UTC
    const flag = e.impact === "HIGH" ? "🔴" : "🟠";
    lines.push(`${flag} ${time} <b>${esc(e.currency)}</b> — ${esc(e.title)}`);
  }

  lines.push(...desk);
  lines.push("", "⚠️ Trades auto-block ±30m around high-impact events.");
  return lines.join("\n");
}

/**
 * Build and send the morning news brief to the configured Telegram chat.
 * No-ops (without throwing) when Telegram is not configured.
 */
export async function sendDailyNewsBrief(): Promise<{ sent: boolean; reason?: string }> {
  if (!isConfigured()) return { sent: false, reason: "telegram_not_configured" };
  const chatId = defaultChatId();
  if (!chatId) return { sent: false, reason: "no_chat_id" };

  const text = await buildNewsBrief();
  const id = await sendMessage(chatId, text);
  return id != null ? { sent: true } : { sent: false, reason: "send_failed" };
}

File: services/backend/app/domain/execution/news_brief.py
"""Morning news brief → Telegram — port of `execution/newsBrief.ts`.

Part of the agent's "check news first" daily routine: before the session,
summarize the high/medium-impact economic events in the next 24h and push them
to the user's Telegram chat.

Best-effort throughout: a missing AI provider, empty calendar, or unconfigured
Telegram must never raise into the scheduler. The same `NewsEvent` rows feed the
risk engine's ±news-window block, so this brief and the trade gate stay in sync.
"""
from __future__ import annotations

import logging
from datetime import datetime, timedelta

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from ...core.serialization import iso
from ...db.enums import Impact
from ...db.models import NewsEvent
from ...integrations.ai import client as ai
from ...integrations.ai.schemas import NewsSummaryRequest
from ...integrations.telegram.client import (
    default_chat_id,
    esc,
    is_configured,
    send_message,
)
from ...jobs.clock import naive_utcnow
from .daily_briefing import collect_market_context

log = logging.getLogger("backend.newsBrief")

DAY = timedelta(days=1)


async def _ai_summary(events: list[NewsEvent]) -> str | None:
    """A short consolidated "what to watch" paragraph, or None on any failure."""
    try:
        headlines = [
            {
                "title": e.title,
                "source": "economic-calendar",
                "publishedAt": iso(e.scheduledAt),
                "body": (
                    f"{e.impact.value}-impact {e.currency} event. "
                    f"Forecast: {e.forecast or 'n/a'}, Previous: {e.previous or 'n/a'}."
                ),
            }
            for e in events[:25]
        ]
        result = await ai.news_summary(
            NewsSummaryRequest.model_validate({"headlines": headlines})
        )
        summary = (result.summary or "").strip()
        return summary or None
    except Exception as exc:  # noqa: BLE001
        log.warning("[newsBrief] AI summary unavailable: %s", exc)
        return None


async def _desk_call_section(session: AsyncSession) -> list[str]:
    """Morning desk call: bias + levels + risks per traded series."""
    contexts = await collect_market_context(session)
    if not contexts:
        return []
    lines = ["", "<b>DESK CALL</b>"]
    for m in contexts:
        dot = "🟢" if m["bias"] == "Bullish" else "🔴" if m["bias"] == "Bearish" else "⚪"
        lines.append(
            f"{dot} <b>{esc(m['symbol'])} {esc(m['timeframe'])}</b> — {esc(m['bias'])}"
        )
        lines.append(esc(m["summary"][:300]))
        if m["keyLevels"]:
            lines.append(f"Levels: {esc(' · '.join(m['keyLevels'][:3]))}")
        if m["risks"]:
            lines.append(f"Risks: {esc(' · '.join(m['risks'][:2]))}")
    return lines


async def build_news_brief(session: AsyncSession, now: datetime | None = None) -> str:
    """The HTML message body for the next-24h news brief."""
    reference = now or naive_utcnow()
    events = (
        (
            await session.execute(
                select(NewsEvent)
                .where(
                    NewsEvent.scheduledAt > reference,
                    NewsEvent.scheduledAt < reference + DAY,
                    NewsEvent.impact.in_([Impact.HIGH, Impact.MEDIUM]),
                )
                .order_by(NewsEvent.scheduledAt.asc())
                .limit(15)
            )
        )
        .scalars()
        .all()
    )

    desk = await _desk_call_section(session)

    if not events:
        return "\n".join(
            [
                "📰 <b>NEWS BRIEF</b> — next 24h (UTC)",
                "",
                "No high- or medium-impact events scheduled. Clear to follow your "
                "plan — still read the chart before any entry.",
                *desk,
            ]
        )

    summary = await _ai_summary(events)
    lines = ["📰 <b>NEWS BRIEF</b> — next 24h (UTC)", ""]
    if summary:
        lines += [esc(summary), ""]

    lines.append("<b>SCHEDULED</b>")
    for e in events:
        time = iso(e.scheduledAt)[11:16]  # HH:MM UTC
        flag = "🔴" if e.impact == Impact.HIGH else "🟠"
        lines.append(f"{flag} {time} <b>{esc(e.currency)}</b> — {esc(e.title)}")

    lines += desk
    lines += ["", "⚠️ Trades auto-block ±30m around high-impact events."]
    return "\n".join(lines)


async def send_daily_news_brief(session: AsyncSession) -> dict[str, object]:
    """Build and send the morning brief. No-ops when Telegram is not configured."""
    if not is_configured():
        return {"sent": False, "reason": "telegram_not_configured"}
    chat_id = default_chat_id()
    if not chat_id:
        return {"sent": False, "reason": "no_chat_id"}

    text = await build_news_brief(session)
    message_id = await send_message(chat_id, text)
    return {"sent": True} if message_id is not None else {"sent": False, "reason": "send_failed"}
